### ISSUE TITLE
Traceback "TypeError: str, bytes or bytearray expected, not NoneType" with non-reachable newsserver

### ISSUE THREAD

**Issue by sanderjo** [sabnzbd/sabnzbd#2345]
git version (and sabnzbdplus 3.7.0 and 3.6.1) gives the traceback below, each time I start SABnzbd (with something in queue).

Probably caused by having www.google.nl as newsserver, which via happyeyeballs does not result in an IP address.

Maybe good to introduce a unit test for this too?

```
2022-11-19 18:04:16,557::INFO::[downloader:672] 1@www.google.nl: Initiating connection
2022-11-19 18:04:19,562::INFO::[happyeyeballs:97] Quickest IP address for www.google.nl (port 563, preferipv6 False) is None
2022-11-19 18:04:19,562::DEBUG::[happyeyeballs:99] Happy Eyeballs lookup and port connect took 3005 ms
2022-11-19 18:04:19,562::DEBUG::[downloader:197] www.google.nl: No successful IP connection was possible
Exception in thread Thread-28:
Traceback (most recent call last):
2022-11-19 18:04:19,575::INFO::[downloader:672] 2@www.google.nl: Initiating connection
  File "/usr/lib/python3.8/threading.py", line 932, in _bootstrap_inner
2022-11-19 18:04:19,575::DEBUG::[downloader:175] www.google.nl: Re-using address None
Exception in thread Thread-29:
2022-11-19 18:04:19,576::INFO::[downloader:672] 3@www.google.nl: Initiating connection
2022-11-19 18:04:19,576::DEBUG::[downloader:175] www.google.nl: Re-using address None
    self.run()
  File "/usr/lib/python3.8/threading.py", line 870, in run
Traceback (most recent call last):
2022-11-19 18:04:19,577::INFO::[downloader:672] 4@www.google.nl: Initiating connection
2022-11-19 18:04:19,577::DEBUG::[downloader:175] www.google.nl: Re-using address None
2022-11-19 18:04:19,577::INFO::[downloader:672] 5@www.google.nl: Initiating connection
2022-11-19 18:04:19,578::DEBUG::[downloader:175] www.google.nl: Re-using address None
  File "/usr/lib/python3.8/threading.py", line 932, in _bootstrap_inner
    self._target(*self._args, **self._kwargs)
2022-11-19 18:04:19,578::INFO::[downloader:672] 6@www.google.nl: Initiating connection
  File "/home/sander/git/sabnzbd/sabnzbd/newswrapper.py", line 340, in connect
2022-11-19 18:04:19,578::DEBUG::[downloader:175] www.google.nl: Re-using address None
Exception in thread Thread-32:
Exception in thread Thread-31:
Traceback (most recent call last):
  File "/usr/lib/python3.8/threading.py", line 932, in _bootstrap_inner
    self.sock.connect((self.host, self.nw.server.port))
TypeError: str, bytes or bytearray expected, not NoneType
Exception in thread Thread-30:
Traceback (most recent call last):
  File "/usr/lib/python3.8/threading.py", line 932, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.8/threading.py", line 870, in run
2022-11-19 18:04:19,580::INFO::[downloader:672] 7@www.google.nl: Initiating connection
    self.run()
  File "/usr/lib/python3.8/threading.py", line 870, in run
2022-11-19 18:04:19,580::DEBUG::[downloader:175] www.google.nl: Re-using address None
    self._target(*self._args, **self._kwargs)
  File "/home/sander/git/sabnzbd/sabnzbd/newswrapper.py", line 340, in connect
    self.run()
    self._target(*self._args, **self._kwargs)
  File "/home/sander/git/sabnzbd/sabnzbd/newswrapper.py", line 340, in connect
    self.sock.connect((self.host, self.nw.server.port))
TypeError: str, bytes or bytearray expected, not NoneType
  File "/usr/lib/python3.8/threading.py", line 870, in run
Traceback (most recent call last):
  File "/usr/lib/python3.8/threading.py", line 932, in _bootstrap_inner
    self._target(*self._args, **self._kwargs)
  File "/home/sander/git/sabnzbd/sabnzbd/newswrapper.py", line 340, in connect
    self.sock.connect((self.host, self.nw.server.port))
TypeError: str, bytes or bytearray expected, not NoneType
2022-11-19 18:04:19,584::INFO::[downloader:672] 8@www.google.nl: Initiating connection
    self.run()
2022-11-19 18:04:19,584::DEBUG::[downloader:175] www.google.nl: Re-using address None
  File "/usr/lib/python3.8/threading.py", line 870, in run
    self._target(*self._args, **self._kwargs)
  File "/home/sander/git/sabnzbd/sabnzbd/newswrapper.py", line 340, in connect
    self.sock.connect((self.host, self.nw.server.port))
TypeError: str, bytes or bytearray expected, not NoneType
Exception in thread Thread-33:
Traceback (most recent call last):
  File "/usr/lib/python3.8/threading.py", line 932, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.8/threading.py", line 870, in run
    self._target(*self._args, **self._kwargs)
  File "/home/sander/git/sabnzbd/sabnzbd/newswrapper.py", line 340, in connect
    self.sock.connect((self.host, self.nw.server.port))
TypeError: str, bytes or bytearray expected, not NoneType
    self.sock.connect((self.host, self.nw.server.port))
TypeError: str, bytes or bytearray expected, not NoneType
Exception in thread Thread-34:
Traceback (most recent call last):
  File "/usr/lib/python3.8/threading.py", line 932, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.8/threading.py", line 870, in run
    self._target(*self._args, **self._kwargs)
  File "/home/sander/git/sabnzbd/sabnzbd/newswrapper.py", line 340, in connect
    self.sock.connect((self.host, self.nw.server.port))
TypeError: str, bytes or bytearray expected, not NoneType
Exception in thread Thread-35:
Traceback (most recent call last):
  File "/usr/lib/python3.8/threading.py", line 932, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.8/threading.py", line 870, in run
    self._target(*self._args, **self._kwargs)
  File "/home/sander/git/sabnzbd/sabnzbd/newswrapper.py", line 340, in connect
    self.sock.connect((self.host, self.nw.server.port))
TypeError: str, bytes or bytearray expected, not NoneType

2022-11-19 18:05:19,662::INFO::[downloader:952] Thread 1@www.google.nl: timed out
2022-11-19 18:05:19,662::DEBUG::[downloader:977] Re-adding article ZpUwNcWlXaVzJeJcBjXmBqCu-1658672640910@nyuu from sometestfile-100MB.part03.rar to server <Server: www.google.nl:563>
2022-11-19 18:05:19,662::INFO::[downloader:952] Thread 2@www.google.nl: timed out
2022-11-19 18:05:19,662::DEBUG::[downloader:977] Re-adding article PoGtRtIyThRwHgJbSrJhXuNq-1658672642009@nyuu from sometestfile-100MB.part04.rar to server <Server: www.google.nl:563>
2022-11-19 18:05:19,662::INFO::[downloader:952] Thread 3@www.google.nl: timed out

```

**Comment by Safihre** [comment]
@puzzledsab I am guessing a side effect from the last PR where we set info to None?

**Comment by sanderjo** [comment]
I have it in 3.6.1 too. But not in GUI, and ... also not in sabnzbd.log (is that even possible?).

**Comment by puzzledsab** [comment]
I'm trying to figure out this issue but I'm not able to recreate the traceback. It doesn't work properly for me either, though, at least with a required server. It will still send the article to the next server after 3 tries. With this patch it will try required servers indefinitely. It will also give a warning about the error the first time it occurs so that the user is aware.

I think the reason why it crashes for @sanderjo is that https://github.com/sabnzbd/sabnzbd/blob/develop/sabnzbd/downloader.py#L173 assumes that if there is busy thread then there is an IP address. I don't know why he doesn't have an IP address at that point, though. It does for me with the latest develop branch.

@sanderjo : The dump gets less messy if you only use 2 connections for the Google server.

**Comment by sanderjo** [comment]
I got your PR (commit confirms that, right?), I started with a fresh ini, filled out the wizard with news.eweka.nl port 8888 (=non-reachable), disabled API-logging (to get cleaner logging), and then I added the 100MB test download, and ... bingo: traceback:

```
sander@zwart2204:~/git/puzzledsab$ ./SABnzbd.py -b1 -f blabla.ini 
2022-11-20 21:59:51,926::INFO::[SABnzbd:1167] --------------------------------
2022-11-20 21:59:51,926::INFO::[SABnzbd:1168] SABnzbd.py-3.8.0-develop
2022-11-20 21:59:51,926::INFO::[misc:1161] [N/A] Running external command: ['git', 'rev-parse', '--short', 'HEAD']
2022-11-20 21:59:51,927::INFO::[SABnzbd:1178] Commit = 1c4a5da69

2022-11-20 21:59:53,255::INFO::[happyeyeballs:97] Quickest IP address for news.eweka.nl (port 8888, preferipv6 False) is None
Exception in thread Thread-25 (connect):
Traceback (most recent call last):
  File "/usr/lib/python3.10/threading.py", line 1016, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.10/threading.py", line 953, in run
    self._target(*self._args, **self._kwargs)
  File "/home/sander/git/puzzledsab/sabnzbd/newswrapper.py", line 340, in connect
    self.sock.connect((self.host, self.nw.server.port))
TypeError: str, bytes or bytearray expected, not NoneType
2022-11-20 22:00:53,320::INFO::[downloader:952] Thread 1@news.eweka.nl: timed out
```

**Comment by sanderjo** [comment]
Yes, looks like it:

```
sander@zwart2204:~$ cat weird_socket_connect.py 
import socket
host = None
port = 8888  # socket server port number

client_socket = socket.socket()  # instantiate
client_socket.connect((host, port))  # connect to the server

```
which gives    
```
sander@zwart2204:~$ python3 weird_socket_connect.py 

Traceback (most recent call last):
  File "/home/sander/weird_socket_connect.py", line 7, in <module>
    client_socket.connect((host, port))  # connect to the server
TypeError: str, bytes or bytearray expected, not NoneType

```

**Comment by sanderjo** [comment]
Ugly hack I tried (but failed): in happyeyeballs I changed the two occurences of `result = None` to `result = ""` .  Result with Commit = 1c4a5da69 and/but orignal sabnzbd.ini (so with good servers defined, at a lower prio): no more tracebacks, but SAB is still not trying the other newservers. :-(

```
2022-11-20 22:21:31,028::INFO::[downloader:672] 1@www.google.nl: Initiating connection
2022-11-20 22:21:34,031::INFO::[happyeyeballs:97] Quickest IP address for www.google.nl (port 119, preferipv6 False) is 
2022-11-20 22:21:34,031::DEBUG::[happyeyeballs:99] Happy Eyeballs lookup and port connect took 3003 ms
2022-11-20 22:21:34,031::DEBUG::[downloader:197] www.google.nl: No successful IP connection was possible
2022-11-20 22:21:34,032::INFO::[notifier:123] Sending notification: Warning - Failed to connect: [Errno 111] Connection refused :119 (type=warning, job_cat=None)
2022-11-20 22:21:34,032::WARNING::[newswrapper:406] Failed to connect: [Errno 111] Connection refused :119
2022-11-20 22:21:35,033::DEBUG::[downloader:977] Re-adding article EkZuHcMrVxSbBbBaSuFgVbMm-1658672638813@nyuu from sometestfile-100MB.part01.rar to server <Server: www.google.nl:119>
2022-11-20 22:22:35,109::INFO::[downloader:672] 1@www.google.nl: Initiating connection
2022-11-20 22:22:38,114::INFO::[happyeyeballs:97] Quickest IP address for www.google.nl (port 119, preferipv6 False) is 
2022-11-20 22:22:38,114::DEBUG::[happyeyeballs:99] Happy Eyeballs lookup and port connect took 3005 ms
2022-11-20 22:22:38,115::DEBUG::[downloader:197] www.google.nl: No successful IP connection was possible
2022-11-20 22:22:38,115::INFO::[newswrapper:404] Failed to connect: [Errno 111] Connection refused :119
2022-11-20 22:22:39,116::DEBUG::[downloader:977] Re-adding article EkZuHcMrVxSbBbBaSuFgVbMm-1658672638813@nyuu from sometestfile-100MB.part01.rar to server <Server: www.google.nl:119>
2022-11-20 22:23:39,188::INFO::[downloader:672] 1@www.google.nl: Initiating connection
2022-11-20 22:23:42,193::INFO::[happyeyeballs:97] Quickest IP address for www.google.nl (port 119, preferipv6 False) is 
2022-11-20 22:23:42,194::DEBUG::[happyeyeballs:99] Happy Eyeballs lookup and port connect took 3005 ms
2022-11-20 22:23:42,194::DEBUG::[downloader:197] www.google.nl: No successful IP connection was possible
2022-11-20 22:23:42,195::INFO::[newswrapper:404] Failed to connect: [Errno 111] Connection refused :119
2022-11-20 22:23:43,196::INFO::[downloader:672] 1@news.eweka.nl: Initiating connection
2022-11-20 22:23:43,222::INFO::[happyeyeballs:97] Quickest IP address for news.eweka.nl (port 563, preferipv6 False) is 81.171.92.205
2022-11-20 22:23:43,222::DEBUG::[happyeyeballs:99] Happy Eyeballs lookup and port connect took 26 ms
2022-11-20 22:23:43,223::DEBUG::[downloader:194] news.eweka.nl: Connecting to address 81.171.92.205
2022-11-20 22:23:43,294::INFO::[newswrapper:346] 1@news.eweka.nl: Connected using TLSv1.3 (TLS_AES_256_GCM_SHA384)
2022-11-20 22:23:44,291::INFO::[downloader:889] Connecting 1@news.eweka.nl finished

(SABnzbd.py:60102): Gtk-CRITICAL **: 22:23:44.993: gtk_widget_get_scale_factor: assertion 'GTK_IS_WIDGET (widget)' failed
2022-11-20 22:24:43,601::INFO::[downloader:672] 1@www.google.nl: Initiating connection
2022-11-20 22:24:46,678::INFO::[happyeyeballs:97] Quickest IP address for www.google.nl (port 119, preferipv6 False) is 
2022-11-20 22:24:46,678::DEBUG::[happyeyeballs:99] Happy Eyeballs lookup and port connect took 3077 ms
2022-11-20 22:24:46,679::DEBUG::[downloader:197] www.google.nl: No successful IP connection was possible
2022-11-20 22:24:46,679::INFO::[newswrapper:404] Failed to connect: [Errno 111] Connection refused :119
2022-11-20 22:24:47,681::DEBUG::[downloader:977] Re-adding article UtJrLhPwPvHrZjTvUjZrCpKw-1658672639558@nyuu from sometestfile-100MB.part02.rar to server <Server: www.google.nl:119>

```

**Comment by puzzledsab** [comment]
We could do the happyeyeballs test first if load_balancing == 2 and then fall back to load_balancing = 0 if it fails to find an IP. It probably wouldn't create a functional IP but at least it would would work like the other ones for error handling and reporting.

**Comment by puzzledsab** [comment]
@sanderjo : Is Required set on the Google server? If so it won't give up. If not it should give up after the number of retries set under Switches / Server / Maximum retries. It will take a long time before you notice any progress with 3 retries for each article.

**Comment by sanderjo** [comment]
No, Required is not on.

After 58 minutes of running, SAB is at 16% of the 100MB. This is that very annoying problem.

![image](https://user-images.githubusercontent.com/1273502/202929566-e30c1695-a3e5-47b4-8d48-14e809b13479.png)


During that 58 minutes, SAB tried google 57 times. And eweka 16 ... in which it must have gotten that 16% of the 100MB test downloade

```
sander@zwart2204:~$ cat .sabnzbd/logs/sabnzbd.log | grep -A10000 "2022-11-20 21:57:55,399" | grep "1@www.google.nl: Initiating connection" | wc
     57     285    4959
sander@zwart2204:~$ cat .sabnzbd/logs/sabnzbd.log | grep -A10000 "2022-11-20 21:57:55,399" | grep "eweka.nl: Initiating connection" | wc
     16      80    1392
```

**Comment by puzzledsab** [comment]
If you set it as optional it will be ignored for 10 minutes after 3 retries (per connection). This part seems to be working as intended.

**Comment by Safihre** [comment]
Closed by #2347
